spinalcordtoolbox/types.py

```diff
@@ -321,7 +321,10 @@ def compute_coordinate_system(self, index):
         return origin, x_prime_axis, y_prime_axis, z_prime_axis, matrix_base, inverse_matrix
 
     def get_projected_coordinates_on_planes(self, coordinates, indexes):
-        return coordinates - multiply(tile(einsum('ij,ij->i', coordinates - self.points[indexes], self.derivatives[indexes]), (3, 1)).transpose(), self.derivatives[indexes])
+        unit_derivatives = self.derivatives[indexes]
+        unit_derivatives /= np.expand_dims(norm(unit_derivatives, axis=1), axis=1)
+        dot_products = np.expand_dims(np.sum((coordinates - self.points[indexes]) * unit_derivatives, axis=1), axis=1)
+        return coordinates - unit_derivatives * dot_products
```

Comment:
Yes, `einsum` is essentially write-only code (aka, unreadable)! I'll try to unpack the old line of code below. But do you mean performance in terms of speed? Or that it's computing the same thing? (It's not actually computing the same thing, because the old code had a math bug in it.)

1. Let's say that the centerline has `n` points in total, and that `len(coordinates) == len(indexes) == k`.
2. `coordinates` is essentially a list of 3D point; it's a numpy array of shape `(k, 3)`.
3. `self.points[indexes]` is a corresponding list of 3D points on the centerline, also a numpy array of shape `(k, 3)`.
4. So, `coordinates - self.points[indexes]` is a list of 3D vectors; for each one, it's a vector pointing from a point on the centerline, to a point in `coordinates`. Shape `(k, 3)` again.
5. Meanwhile, `self.derivatives[indexes]` is another list of 3D vectors; each vector points in a direction parallel to the centerline at the corresponding point of `self.points[indexes]`. Shape `(k, 3)` yet again.
6. The `einsum` call is directly equivalent to this nested loop:

   ```py
   input_1 = coordinates - self.points[indexes]
   input_2 = self.derivatives[indexes]
   result = np.zeros(k)  # a 1D array of length k
   for i in range(k):
       for j in range(3):
           result[i] += input_1[i, j] * input_2[i, j]
   ```

   So, this is computing `k` dot products, for the `k` vectors in `coordinates - self.points[indexes]` together with the corresponding vectors in `self.derivatives[indexes]`.
7. The call to `tile(result, (3, 1))` is turning the numpy array of length `k` into an array of shape `(3, k)` repeating each of the `k` entries 3 times. Then `.transpose()` flips it into shape `(k, 3)`, to match all the other arrays we have. (Idiomatic numpy usage would be to use reshaping and broadcasting instead of `tile`, I think.)
8. Finally, `np.multiply` is the usual numpy operation of entry-wise multiplication of two arrays of shape `(k, 3)`, but not written as `*` for some reason. Mathematically, the result is a list of `k` vectors, obtained by scaling each vector in `self.derivatives[indexes]` by a certain dot product.
9. Finally finally, the result is subtracted (again, for each of the `k` indices) from the original points in `coordinates`.
10. If the vectors in `self.derivatives[indexes]` were unit-length (but they are not, this is the bug), then this would be the right linear algebra formula for projecting each vector in `coordinates` onto the plane perpendicular to the spinal cord at the corresponding point in `self.points[indexes]`. That's why I added the `/=` step for `unit_derivatives`.

The performance of the new code should be pretty similar to the old code; it's not significantly changing the number of operations, the nesting of loops, or the sizes of the arrays involved.

Reply:
> Or that it's computing the same thing? (It's not actually computing the same thing, because the old code had a math bug in it.)

This exactly! 

Thank you so much for unpacking everything. Woof, there's a lot of complexity in one line of code!!